egressgw: steer traffic to the right interface using BPF

jibi opened the pull request:
see commits

Fixes: https://github.com/cilium/cilium/issues/23504

```release-note
EgressGW: interface selection is now done with BPF, using --install-egress-gateway-routes is no longer needed.
```

julianwiedmann commented on bpf/lib/nat.h:
```diff
@@ -572,6 +575,12 @@ static __always_inline int snat_v4_rewrite_egress(struct __ctx_buff *ctx,
+	if (target && target->egress_gateway)
+		return egress_gw_fib_lookup_and_redirect(ctx, state->to_saddr);
+#endif
+
```
Layering-wise, I think we should lift this into `nodeport_snat_fwd_ipv4()`. The `target` should have all the info we need from there (`target.addr == state->to_saddr`, right?)

And `tail_handle_snat_fwd_ipv4()` needs a check to only `send_trace_notify()` for `CTX_ACT_OK`.

jibi replied:
> Layering-wise, I think we should lift this into `nodeport_snat_fwd_ipv4()`. The `target` should have all the info we need from there (`target.addr == state->to_saddr`, right?)

yep, I was a bit worried about that stray `send_drop_notify_ext` in lib/egress_policies.h :hmmyes:
now drop notifications should get handled by the `tail_handle_snat_fwd_ipv4` caller

> And `tail_handle_snat_fwd_ipv4()` needs a check to only `send_trace_notify()` for `CTX_ACT_OK`.

:bulb:

jibi committed:
egressgw: mark --install-egress-gateway-routes as deprecated

this flag has now no effect and will be removed in v1.16

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: egressgw: add support for fib_lookup

when a packet matching an egress gateway policy needs to leave the node,
it can do so through the wrong interface. This can happen when the
interface with the route for that packet is different from the one with
the egress IP assigned to.

The current solution to this is to install additional IP rules and
routes that match egress gateway traffic and steer it to the correct
interface with the egress IP. The drawback of this approach is that
installing and maintaining these rules is getting more and more complex
and consumes resources on the host.

This commit introduces a new approach to solve this, based on
fib lookups and redirects in the datapath. In practice, after a
packet matching an egress gateway policy has been masqueraded, we run a
fib lookup on the packet with the new source (egress) IP, and in case
the packet is redirected to the correct interface with the egress IP
assigned to.

Fixes: #23504
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: bpf_host: skip host firewall in cil_to_netdev if packet is SNATed

when egress gateway traffic gets masqueraded, it can get redirected to a
different interface that is also running bpf_host.

When this happens, the host firewall logic should be skipped as it would
be incorrect to enforce again any policy: after being masqueraded and
redirected, the original pod identity is lost, and the host firewall
would incorrectly identify this traffic as host or world

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
egressgw: remove IP rules/and routes logic

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: skip tail_handle_snat_fwd_ipv4 trace event if packet was redirected

Don't emit a trace event if the packet has been redirected to another
interface. This can happen for egress gateway traffic that needs to
egress from the interface to which the egress IP is assigned to.

Suggested-by: Julian Wiedmann <jwi@isovalent.com>
Signed-off-by: Gilberto Bertin <jibi@cilium.io>

lmb commented on bpf/lib/egress_gateway.h:
```diff
@@ -137,5 +139,22 @@ bool egress_gw_reply_needs_redirect(struct iphdr *ip4 __maybe_unused,
+	if (old_oif == fib_params.l.ifindex)
+		return CTX_ACT_OK;
+
+	return fib_do_redirect(ctx, true, &fib_params, ext_err, (int *)&old_oif);
```
Why is this an `int` here?

jibi replied:
`fib_do_redirect()` was introduced in https://github.com/cilium/cilium/pull/26136, so I'll defer that to @ldelossa as I'm not quite sure about the context (especially given that in that function, `oif` is always used as a u32)

julianwiedmann replied:
We've been going back and forth on `u32` and `int`, depending on who is the last to touch the relevant FIB / redirect code :slightly_smiling_face:. I believe it originates from `ctx_redirect()` taking an int.

lmb commented on bpf/lib/nodeport.h:
```diff
@@ -3055,7 +3060,8 @@ int tail_handle_snat_fwd_ipv4(struct __ctx_buff *ctx)
 						  CTX_ACT_DROP, METRIC_EGRESS);
 
-	send_trace_notify(ctx, obs_point, 0, 0, 0, 0, trace.reason, trace.monitor);
+	if (ret == CTX_ACT_OK)
```
This could use a comment. I'm guessing that nodeport_snat_fwd_ipv4 can now also redirect in addition to just rewriting headers?

jibi replied:
yeah it's just in the commit message for now, I'll add a comment also here

julianwiedmann replied:
Just a nit, but maybe adding at least a comment in the corresponding IPv6 path would be good (explaining why this handling is *not* needed there at the moment).

lmb commented on pkg/egressgateway/manager_privileged_test.go:
```diff
@@ -51,13 +50,13 @@ const (
 	excludedCIDR1   = "1.1.1.22/32"
 	excludedCIDR2   = "1.1.1.240/30"
 
+	zeroIP4 = "0.0.0.0"
```
Nit: gratuitous change?

lmb commented on pkg/datapath/linux/linux_defaults/linux_defaults.go:
```diff
@@ -19,13 +19,6 @@ const (
-	// Each ENI interface will have its own table starting with this offset. It
-	// is 300 because it is highly unlikely to collide with the main routing
-	// table which is between 253-255. See ip-route(8).
-	RouteTableEgressGatewayInterfacesOffset = 300
```
What abut `RulePriorityEgressGateway`?

jibi replied:
missed it

lmb reviewed (commented):
I am very very very happy about this change. I don't know enough about the datapath to know whether the logic is sound though. Do we need someone in addition to Julian to review?

brb reviewed (approved):
Reviewed `bpf/` - nice work! Less moving parts FTW.

BTW, when doing upgrade from v1.14 -> v1.15, shouldn't we remove old IP rules/routes?

julianwiedmann commented on bpf/lib/egress_gateway.h:
```diff
@@ -22,6 +24,23 @@
+	struct bpf_fib_lookup_padded fib_params = {};
+	__u32 old_oif = ctx_get_ifindex(ctx);
+
+	*ext_err = (__s8)fib_lookup_v4(ctx, &fib_params, egress_ip, 0, 0);
```
Could you explain why it's ok to set `.daddr` to 0 here? I would have expected the actual cluster-external destination.

jibi replied:
I thought it was ok to set just the egress/source IP as we want to identify the interface with that address assigned to (and that should be independent from the destination), or do you have a specific case in mind where this would not not work?

julianwiedmann replied:
Hmm not in combination with how we do EgressGW. We already say that all selected traffic should exit through *one* specific interface.

I'm just a bit wary about having *one* caller in the codebase that doesn't set the `daddr`. That makes it easy to miss bugs / corner cases, because no other path would be exposed to it.

jibi replied:
> I'm just a bit wary about having one caller in the codebase that doesn't set the daddr. That makes it easy to miss bugs / corner cases, because no other path would be exposed to it.

yeah good point, I'll just pass `tuple.daddr` as well :+1:

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -1361,6 +1361,9 @@ int cil_to_netdev(struct __ctx_buff *ctx __maybe_unused)
 #endif
 
 #ifdef ENABLE_HOST_FIREWALL
+	if (ctx_snat_done(ctx))
+		goto skip_host_firewall;
+
```
The reasoning makes sense, although the scope of the change is a bit wider (it also affects eg. NAT-forwarded LB connections in TC, and RevDNATed replies, ...).

Using the SNAT-done flag for such policy decisions feels a bit fragile. I would hope that we can come up with a better solution over time. But good enough for me atm :+1:.

julianwiedmann commented on pkg/egressgateway/manager.go:
```diff
@@ -87,6 +87,7 @@ var defaultConfig = Config{
 
 func (def Config) Flags(flags *pflag.FlagSet) {
 	flags.Bool("install-egress-gateway-routes", def.InstallEgressGatewayRoutes, "Install egress gateway IP rules and routes in order to properly steer egress gateway traffic to the correct ENI interface")
+	flags.MarkDeprecated("install-egress-gateway-routes", "This option is deprecated, has no effect, and will be removed in v1.16")
```
I remember at some point we said that the "FIB lookup" logic would only get enabled iff this option is set. To ensure that the egress interface would have bpf_host attached. Did you come up with another solution, or do we just want to take the simple approach first?

jibi replied:
I'm just a bit eager to rip out this code (as otherwise it will stay around for yet another stable release) and start testing the feature with fib lookups always enabled

> To ensure that the egress interface would have bpf_host attached

I thought we concluded that these interfaces will have already bpf_host attached anyway (otherwise it's not possible to reverse translate reply traffic) so that shouldn't be an issue

julianwiedmann replied:
> I thought we concluded that these interfaces will have already bpf_host attached anyway (otherwise it's not possible to reverse translate reply traffic) so that shouldn't be an issue

Ah yes that sounds reasonable (and I probably just don't remember discussing it before :disappointed:). Agree with the sentiment to roll this out ASAP.

julianwiedmann reviewed (commented):
Just a few smaller questions for clarification, overall lgtm :rocket:.

jibi committed:
egressgw: clean up stale IP rules and routes

this commit adds logic to ensure the manager cleans up any stale IP
rules and routes from previous Cilium versions which still use those to
steer egress gateway traffic to the correct interface

Signed-off-by: Gilberto Bertin <jibi@cilium.io>
